README

                                    GNUitar
                           Guitar processor software

                               What is GNUitar ?

   GNUitar is a real-time sound effects software that allows you to use
   your computer as a guitar processor. It has GTK+ based interface. It
   can be compiled on any flavor of UNIX that have GTK+ 1.2, 2.0, Glib,
   pthreads and OSS sound driver. It also works on Windows. This program
   was inspired by two earlier works:
     * Ele 0.1 by Morris Slutsky
       [1]http://www-personal.umich.edu/~mslutsky/elepage/index.html
     * Guitar FX Processor by Marin Vlah
       [2]http://fly.srk.fer.hr/~mvlah/fx_processor.html

   GNUitar is free software and is distributed under GNU GPL license.

   GNUitar is pronounced like "GNU Guitar". English is not my native
   language, so when I invented this word I did not think how it should
   be pronounced. Unless someone has a better idea, let's pronounce it
   like this for now.

                                   Download

   [3]http://www.gnuitar.com/downloads.php
   [4]http://freshmeat.net/projects/gnuitar
   [5]http://sourceforge.net/projects/gnuitar

                                 Requirements

   You will need:
     * GTK+ versions 1.2.6 or better, or GTK+ 2.0 or better
     * GLIB 1.2 or better
     * POSIX threads on UNIX
     * full-duplex sound card
     * To compile: GCC on UNIX, Microsoft Visual C++ 6.0 or 7.0 on
       Windows

                                  Performance

   GNUitar should run on even low-end hardware. It has been reported to
   work on even a Pentium II / 300 MHz computer. However, some of the
   effects demand more CPU power, and probably ~1 GHz machine is required
   for running complicated effect setups.

   When running the program, make sure you close all unnecessary
   applications, to free more resources. Avoid anything that can cause
   lots of disk I/O or sudden bursts of CPU usage.

   GNUitar runs with increased priority if available; this may cause
   hang-ups and delayed system response on low-end machines potentially
   making the computer unusable. Therefore you should take care to save
   all important data in other applications before launching GNUitar.
   This situation probably does not occur in practice, but with elevated
   task priorities the consequences for misbehaviour become greater.

Performance - UNIX notes

   GNUitar executable file should be setuid root to process sound using
   real-time priority class; otherwise you'll hear glitches and delays
   while switching between windows. The program drops privileges as soon
   as the effect-processing thread has been started, even before any GTK+
   initialization or command line handling, so it should be safe on any
   system. GNUitar has a latency of about 4-8ms on Linux/Pentium
   II/450/AWE64 ISA!

Performance - Windows notes

   Windows have complex problems with latency when processing audio in
   real time, mostly because of its non-uniform drivers architecture.
   GNUitar latency on Windows is much higher than on Linux. The real
   latency depends mostly on the sound card's driver.

   GNUitar for Windows has two playback methods:
     * Playback via MME
     * Playback via DirectSound

   The first method uses the standard MME API (functions waveOut*()). The
   second method requires DirectX to be installed and uses DirectSound
   for playback. The method can be switched from the sampling parameters
   dialog, by checking/unchecking the "Playback via DirectSound"
   checkbox.

   DirectSound playback shows the best results that are almost close to
   Linux performance. The MME playback exists for compatibility purposes,
   and for another reason. If a driver is not optimized for DirectSound,
   Windows will automatically emulate DirectSound output using the MME
   devices. If a WDM driver is used (see below), DirectSound support is
   not implemented by the driver developer but by the operating system.
   MME playback support exists specially for this case, when the output
   via DirectSound is emulated by OS.

  Windows Sound Drivers Overview

   There are few kinds of sound drivers for Windows: old VXD (Win95/98),
   NT4-kernel style, and modern WDM drivers that were introduced in
   Windows 98/SE. AFAIK WDM drivers provide much lower latency.
   Therefore, avoid VXD drivers; use modern WDM drivers instead, if
   possible.

   The difference in latency between two kinds of drivers is really
   noticeable: I had 100ms up to ~400ms on Pentium III/850/ISA
   AWE64/VXD/MME playback, and ~60ms on Pentium MMX/166/Yamaha
   OPL3/WDM/MME playback laptop. Try to start/stop sound few times, if
   the initial latency is bad.

   The kind of bus (PCI Express/PCI/ISA) of the sound card does not
   affect the latency that much, it's most often a software issue. Old
   sound cards that are unusable in Windows are likely to work just fine
   in Linux.

How To Control Latency

The latency can be controlled to a high degree in the Options dialog. It can be
invoked from the Options->Sampling Parameters menu (or by Ctrl-P). You will
need to stop playback before entering this dialog. There is an option called
"Fragment Size". The lower the fragment size is, the better is the latency, but
this is not a free lunch. If the latency is too low, the CPU drain from
constantly attending to the sound card leaves no time for anything else! You
may need to increase the fragment size on low-end computers. Try increasing
this value if you hear repeated scratches from the speakers.

General notes on how to achieve the best performance:

     * The lower the sampling rate, the less there is CPU load. Drawback
       is the lower sound quality.
     * The lower the fragment size, the better is the latency. Drawback
       is the higher system load.
     * The audible periodic scratches (DirectSound output) can be fixed
       by decreasing the overrun threshold (sampling parameters dialog).
     * Increase the fragment size and decrease the sampling rate on
       low-end CPU, to gain the best latency/overruns/load ratio.
     * On Windows, prefer WDM drivers, if possible. Try both MME and
       DirectSound playback; choose which is the best.
     * Prefer Linux over Windows. Properly tuned Linux kernel has 10-100
       times better latency on the same hardware.
     * Make gnuitar executable setuid root on UNIX, to allow it run with
       increased priority.

                                 Installation

   See [6]INSTALL file for common installation notes.

                            Interface and Controls

   There are 3 areas in the main window. The right area is a list of all
   available effects. The central area contains effects that are
   currently used. There are few buttons right to it should be used to
   add/remove effects and change its order. Each effect has separate
   top-level control window with appropriate sliders. Each effect-control
   box is shown in the window manager task bar.

   The left area contains available effect layouts, or presets, and
   button to add the one. Layout is a "snapshot" of your effects and its'
   settings, you can load/save using "File" menu.

   Big "Switch" button is used to switch layouts. In this manner, you can
   change current sound by one mouse/keyboard click.

   Big "START/STOP" button is used to start/stop playback. You may want
   to try pressing it a few times if you experience buffer overruns or
   broken sound output.

   You can write track of what you play to a file. Just click check-box
   "Write track" at the bottom of program window, enter file name and
   play. Don't forget that continuous track write can fill out your hard
   drive. The track file format is WAV data on Windows and Linux.

   Linux WAV writing depends on libsndfile. However, if you do not have
   sndfile, then the data is raw 16-bit signed data, with the sampling
   rate and channels as defined in the sampling parameters dialog. You
   can convert it with [7]sox like this:
    sox -w -s -c 1 -r 44100 track.raw track.wav,

   and then to [8]ogg vorbis:
    oggenc track.wav

                                    Effects

   The controls and description of the effects follow below.

Autowah

   The autowah is implemented using a digital model of the moog ladder
   filter. The model was designed by Antti Huovilainen, and it was
   published at DAFx'04. On low resonance values it acts a lot like a 2nd
   order bandpass filter with some saturation colouring the sound into a
   warm, bloomy sound. On high resonance values, the wah becomes a sound
   synthesizer itself.

   period
          length of one cycle of effect

   low freq
          lowest frequency (the cycle ends at this freq.)

   high freq
          high frequency (the cycle starts at this freq.)

   dry/wet
          mix clean sound with processed sound

   resonance
          choose resonance of the moog ladder filter. Values above 110 %
          lead to self-supporting oscillations.

   continuous sweep
          The Wah can be run in two modes, one where it sweeps between
          the low and high frequency in /\/\/\/ type manner, and another
          where it "listens" to input signal and tries to detect guitar
          picks for triggering the wah sweep synchronously with guitar
          picking. The pick detection works best if the wah receives
          unaltered input signal from the guitar, rather than say
          distorted input from the overdrive effect.

Distortion

   distort
          power of effect

   level
          volume in case you need to mute it quickly

   saturation
          high-frequency "sand" in the sound

   lowpass
          lowpass filter that can be used to change the sound feel

Delay

   A no-frills time-delay effect that can also split original signal in
   several physical output channels if available.

   decay
          decay % between subsequent repeats

   time
          delay time between repeats

   repeat
          number of repeats

   multichannel
          (if available) switch effect to circulate delay voices through
          available channels

Reverberator

   Reverberator produces echoed "space" sound. Controls:

   delay
          delay before repeating starts

   wet
          "wet" (processed) sound volume

   dry
          "dry" (clean) sound volume

   regen
          decay % between subsequent repeats

Tremolo

   Tremolo is a pitch shifter that does not preserve note relationships
   (in other words, a bit like a genuine tremolo bar). This effect is
   really a version of Stereo phaser with Wet at 100 % and only one
   channel used for output.

   period
          length of one cycle of effect

   amplitude
          depth of modulation

Chorus / Flanger

   This effect mixes a time-delayed versions of input with the original
   signal simulating several players playing the same tune. Variations of
   this concept also produce flanger and vibrato effects.

   ChorusUse several voices and long base delay time.

   FlangerWith just 1 voice and low delay time, the effect becomes a
   flanger. However, by adding more voices and increasing the delay, the
   sound again turns choruslike. If the multichannel mode is available
   and enabled, the effect distributes one voice per output channel.

   Vibrato If dry/wet at 100% (fully wet) and 1 voice, this effect is a
   vibrato.

   delay
          fixed time delay before mixing

   depth
          the depth of modulation

   period
          the length of modulation cycle

   voices
          number of chorus voices to mix

   dry/wet
          dry vs. wet mixing ratio

   regen
          feedbacks some of the output back to input

Echo

   Another reverberation effect, but not like the others. Echo simulates
   a large hall echo using varying length of delays in all its voices,
   and produces complicated echo pattern that simulates large hall full
   of hard surfaces of different sizes.

   decay
          decay % between subsequent repeats

   voices
          number of independent echo voices

   delay
          sets the length of maximum echo delay

Phaser

   Phaser operates by introducing a variable frequency dependant delay
   over the audio signal. If the dry/wet ratio is at 50 %, the delayed
   components will produce interference with the original signal which
   produces moving notches across the audio spectrum, producing the
   phaser sound. However, if dry/wet ratio is at 100%, this effect
   becomes a vibrato.

   period
          length of the modulation cycle

   depth
          an unitless measure of the maximum delay imposed by the effect
          on the audio

   dry/wet
          the dry vs. wet mixing ratio

Rotary speaker

   Rotary speaker is an implementation of the combined horn and bass
   speaker mounted on opposite sides of a rotating cabinet. As the
   cabinet turns, the doppler effect shifts the audio spectrum up and
   down, depending on whether the speaker is receding or approaching.

   period
          Time for one complete rotation of the cabinet.

Vibrato

   Slows down and speeds up playback to achieve an effect like old tape.
   This is done by time-delaying the playback in a long sinuous curve.
     * period - length of one cycle of effect
     * amplitude - maximum time delay

Overdrive

   Another cool distortion. Emulates the distortion circuits in many
   popular overdrive pedals such as Ibanez TS9.

   drive
          how much amplification

   clip
          changes tone from TS9-type fuzz to less powerful distortion

   treble
          tone control

Tube amplifier

   Because Overdrive is limited to simulating the Ibanez TS9, another
   effect is required for more powerful distortion sounds. The effect
   loosely models a tube amplifier with a variable number of output
   stages, each which add different kind of clipping to the sound. Some
   tone controls are provided for additional flexibility.

   stages
          Number of stages. All can be used to produce distortion, 5
          stages being very close to overkill.

   gain
          Gain per stage. Higher gains drive the later stages into more
          clipping.

   bass cut
          A highpass filter with tunable cutoff frequency. Increasing
          this value reduces bass from the output.

   middle cut
          Decreasing the value of the control introduces a deepening
          notch at 720 Hz, allowing the effect to be used for dark
          distortion sounds.

   treble cut
          A lowpass filter with tunable cutoff frequency. This can be
          used to change the tone and control noise.

Noise gate

   Simple noise reduction effect.

  IMPORTANT NOTES

     * Should be used BEFORE the distortion.
     * If you want to achieve a note slight attack (fade in), you must
       increase the Hold time as much as possible. Otherwise with the low
       Attack time and low signal you the effect will be constantly false
       triggered, therefore breaking the fade-in effect

   Controls are:
     * threshold - minimal volume of the sound required to pass signal to
       the output.
     * hold - if the signal is below threshold during this time, it will
       be muted. Should be as low as possible. Simply - it is an effect
       triggering time.
     * release time - if the signal is below threshold, the playback is
       not muted immediately but fades out this time instead
     * attack - if the signal is above the threshold, it will fade in
       this time. Usually should be 0, but having it non-zero will
       produce interesting effect just like the violin sound.
     * hysteresis - the threshold required to turn off the playback when
       it is already on (the regular threshold affects only the growing
       signal, while hysteresis affects fading signal). Should not be
       made larger than the threshold value.

Equalizer bank

   A no-frills equalizer effect with plenty of sliders for fine grained
   control of the frequency response. Equalizers are rumoured to be
   especially useful before and after distortion effects.

Pitch shifter

   Pitch shifter alters audible pitch by dividing the input into several
   small buffers and playing through them in a different rate, causing
   the pitch alteration. Because rate shifting implies generating new
   sample data or omitting some of the sample data, there is a
   discontinuity problem at buffer edges which is handled through
   weighting the buffer data with windowing function and overlapping the
   buffers in such a way that sample amplitude remains fairly constant.

   Bugs With many pitch shifting values several frequencies encounter
   difficult phase cancellation that may render some notes almost
   inaudible or significantly alter their characteristics. Additionally,
   as the new sample data is "generated" from audio data across several
   hundred milliseconds, there will be echo effects when adjusting the
   rate upwards. Finally, the dry signal is delayed by this effect to
   control some of the phase cancellation during wet/dry mixing phase.
   The latency introduced depends on the amount of rate shifting.

   halfnote
          how many half-notes to adjust upwards or downwards

   finetune
          adjust for in-between tones for half-notes

   wet/dry
          select mix % between unprocessed and processed signal

Tuner

   This is a simple general-purpose instrument tuner. The display
   contains:
     * the current detected frequency of the sound
     * the ideal frequency of the sound
     * the note closest to the measured frequency
     * an indicator that shows how the two frequencies are matched
     * list of typical guitar tunings, and indicators that light up when
       frequency matches their pretabulated frequencies.

   How to tune

   In nutshell: if the indicator veers to left, tune up; if it veers to
   right, tune down.

   Tuner is generally agnostic to the desired tuning. Some indicators are
   provided that match some of the most typical guitar tunings, but

NEWS
Version 0.6.0
Effects:
- Distortion 2 reworked and renamed to Overdrive;
- Tuner with layouts (6-string guitar, 4-str. bass etc; easily extendable)
- pitch shift effect;
- built-in bias elimination and noise reduction;
- reworked much more realistic echo;
- reworked chorus - changed controls;
- reworked phaser - the previous effect was actually a wah modification rather
  than phaser;
- tremolo and vibrato names swapped;
- tremolo algo changed;
- reworked auto-wah with two modes -- sweep constantly,
  or sweep only the note playing;
- tubeamp effect -- it picks up the things where the overdrive had left it;
- simple amp effect;
- rotary speaker effect;
Internals:
- ALSA support;
- JACK support;
- added infrastructure to support multiple sound drivers. Currently supported
  ones are OSS, ALSA (*NIX), MMSystem, DirectSound (Win32).
- added background for building stereo/quadro effects;
- added background for having different number of input and output channels,
  if the audio driver supports such;
- unified lookup tables (previously there were many ways of calculating sin() -
  via lookup table, and via math.h's sin());
- unified "backbuf" memory storage for memory based effects (echo, delay etc)
- many misc. improvements in effects;
Interface:
- simple vU meter (based on GTK progress bar);
- simple input and output volume control;
- added configuration .ini-style file in home directory .gnuitarrc;
- presets are stored in text .ini-style format;
- effect windows can be closed by clicking on [x] icon and thus removed
  from active effects list (more intuitive than before);
- save track to .wav on *NIX using libsndfile library;
- suggest a date/time-based filename when writing a track;
Misc:
- added some optimizations on Intel systems -- gcc -msse{1|2|3};
- simplistic debian package;

5 May 2005, Version 0.3.2
- some minor cleanups

26 Apr 2005, Version 0.3.2-pre2
- fixed the help contents bug on win2k
- fixed the 100% cpu load bug with zeroes in the distort2 input
- fixed autowah on windows

18 Apr 2005, Version 0.3.2-pre1
- re-worked distort2 effect; no need to download few Mbytes of lookup tables
- equalizer effect
- GTK 2 port
- updated docs

18 April 2003, Version 0.3.1
- New distortion effect
- Noise gate effect

28 March 2003, Version 0.3.0
- DirectSound playback on Windows DRAMATICALLY improves the latency
- Tooltips added for most major controls.

13 March 2003, Version 0.2.1
- Change sampling params from the interface;
- Meaningful measure units (%, ms) for every slider control;
- New phasor option - turn on bandpass filter;
- A lot of code cleanups;
- Delay was fixed to reset memory on "Time" slider update;
- Minor documentation updates.

6 February 2003, Version 0.2.0
- Windows port;
- Write track to .wav file in Win32 version;
- Meaningful measure units for vibrato and reverb effects (ms and percents);
- HTML documentation; launch browser from a program to see the doc;
- Tremolo bugfix: certain values of the effect params caused array bound
  error and segfault;
- Added icon ;-)
- Major code cleanups for easier maintenance.

25 March 2002, Version 0.1.0-beta 3
- removed clip function from distortion - it sounds pretty bad;
- Bank clist widget width auto-resizing;
- effect control pop-ups cannot be closed from window manager;
- updated doc.

11 March 2002, Version 0.1.0-beta 2
- bugfix with main window destroy code;
- restrict to overwrite existing files by track_write() when in setuid root mode.

11 January 2001, Version 0.1.0-beta 1
- Initial release.


src/main.c
/*
 * GNUitar
 * Main module
 * Copyright (C) 2000,2001,2003 Max Rudensky         <fonin@ziet.zhitomir.ua>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id$
 *
 * $Log$
 * Revision 1.55  2006/05/25 09:03:05  alankila
 * - replace the SSE code with even faster version. Tubeamp effect now runs
 *   20 % faster on my computer. Add some alignment directives to make future
 *   use of SSE potentially easier.
 *
 * Revision 1.54  2006/05/20 09:56:58  alankila
 * - move audio_driver_str and audio_driver_enabled into driver structure
 * - Win32 drivers are ugly, with the need to differentiate between
 *   DirectX and MMSystem operation through dsound variable. The driver
 *   should probably be split with dsound-specific parts in its own driver.
 *
 * Revision 1.53  2006/05/19 15:12:54  alankila
 * I keep on getting rattles with ALSA playback, seems like ALSA doesn't
 * know when to swap buffers or allows write to go on too easily. I
 * performed a major overhaul/cleanup in trying to kill this bug for good.
 *
 * - fix confusion about what "buffer_size" really means and how many bytes
 *   it takes.
 *   - buffer size is ALWAYS the fragment size in all audio drivers
 *     (ALSA, OSS, Win32 driver)
 *   - it follows that memory needed is buffer_size times maximum
 *     frame size. (32-bit samples, 4 channels, max buffer size.)
 *   - latency calculation updated, but it may be incorrect yet
 * - add write buffer for faster ALSA read-write cycle. (Hopefully this
 *   reduces buffer underruns and rattles and all sort of ugliness we
 *   have with ALSA)
 * - redesign the ALSA configuration code. Now we let ALSA choose the
 *   parameters during the adjustment phase, then we try same for playback.
 * - some bugs squashed in relation to this, variables renamed, etc.
 * - if opening audio driver fails, do not kill the user's audio_driver
 *   choice. (It makes configuring bits, channels, etc. difficult.)
 *   We try to track whether processing is on/off through new variable,
 *   audio_driver_enabled.
 *
 * Note: all the GUI code related to audio is in need of a major overhaul.
 * Several variables should be renamed, variable visibility better controlled.
 *
 * Revision 1.52  2006/05/01 10:23:54  anarsoul
 * Alsa device is selectable and input volume is adjustable now. Added new filter - amp.
 *
 * Revision 1.50  2005/09/28 19:55:04  fonin
 * DirectSound hook on startup
 *
 * Revision 1.49  2005/09/12 22:10:50  alankila
 * - add missing checks into configure.in to silence autoscan's warnings
 *
 * Revision 1.48  2005/09/12 09:42:25  fonin
 * - MSVC compatibility fixes
 *
 * Revision 1.47  2005/09/04 23:28:32  alankila
 * - in case audio driver opening fails, set driver to invalid
 *
 * Revision 1.46  2005/09/04 21:04:21  alankila
 * - handle no audio driver more gracefully
 *
 * Revision 1.45  2005/09/04 20:45:01  alankila
 * - store audio driver into config
 * - make it possible to start gnuitar with invalid audio driver and enter
 *   options and correct the driver. Some rough edges still remain with
 *   the start/stop button, mutexes, etc.
 *
 * Revision 1.44  2005/09/04 16:06:59  alankila
 * - first multichannel effect: delay
 * - need to use surround40 driver in alsa
 * - introduce new buffer data_swap so that effects need not reserve buffers
 * - correct off-by-one error in multichannel adapting
 *
 * Revision 1.43  2005/09/04 01:51:09  alankila
 * - GKeyFile-based preset load/save
 * - still need locale-immune %lf for printf and sscanf
 *
 * Revision 1.42  2005/09/03 20:20:42  alankila
 * - create audio_driver type and write all the driver stuff into it. This
 *   faciliates carrying configuration data about the capabilities of
 *   a specific audio driver and uses less global variables.
 *
 * Revision 1.41  2005/08/28 21:41:28  fonin
 * Portability: introduced new functions for mutexes
 *
 * Revision 1.40  2005/08/28 14:04:04  alankila
 * - OSS copypaste error fix
 * - remove my_log2 in favour of doing pow, trunc, log.
 * - OSS driver rounds buffer sizes to suitable values by itself now. There's
 *   a precedent in tuning user parameters automatically in ALSA code. The
 *   new behaviour rounds buffer size down, though.
 *
 * Revision 1.39  2005/08/28 12:39:01  alankila
 * - make audio_lock a real mutex
 * - fix mutex cleanup at exit
 *
 * Revision 1.38  2005/08/28 12:28:44  alankila
 * switch to GMutex that is also available on win32
 *
 * Revision 1.37  2005/08/27 19:05:43  alankila
 * - introduce SAMPLE16 and SAMPLE32 types, and switch
 *
 * Revision 1.36  2005/08/27 18:11:35  alankila
 * - support 32-bit sampling
 * - use 24-bit precision in integer arithmetics
 * - fix effects that contain assumptions about absolute sample values
 *
 * Revision 1.35  2005/08/26 15:59:56  fonin
 * Audio driver now can be chosen by user
 *
 * Revision 1.34  2005/08/25 19:52:11  fonin
 * Removed code for demo version
 *
 * Revision 1.33  2005/08/24 22:33:02  alankila
 * - avoid reopening sound device at exit in order to cleanly shut it down the
 *   next moment
 *
 * Revision 1.32  2005/08/24 21:59:00  alankila
 * some minor cleanup yet
 *
 * Revision 1.31  2005/08/24 21:44:44  alankila
 * - split sound drivers off main.c
 * - add support for alsa
 * - rework thread locking
 * - in this version, sound drivers are chosen at compile time
 * - windows driver is probably broken
 *
 * Revision 1.30  2005/08/24 18:41:36  fonin
 * Bugfix: when the # of samples read != buffer size, program exited. This condition sometimes happen when press on STOP button.
 *
 * Revision 1.29  2005/08/22 22:11:59  alankila
 * - change RC filters to accept data_block
 * - LC filters have no concept of "LOWPASS" or "HIGHPASS" filtering, there's
 *   just filter_no.
 * - remove unused SAMPLE8 typedef
 *
 * Revision 1.28  2005/08/22 11:07:27  alankila
 * - move last bits of tracker support off main.c to pump.c
 * - add settings loader/saver for GTK2, now needs GTK+ 2.6 in minimum
 *
 * Revision 1.27  2005/08/21 23:44:13  alankila
 * - use libsndfile on Linux to write audio as .wav
 * - move clipping tests into pump.c to save writing it in tracker and 3 times
 *   in main.c
 * - give default name to .wav from current date and time (in ISO format)
 * - there's a weird bug if you cancel the file dialog, it pops up again!
 *   I have to look into what's going on.
 *
 * Revision 1.26  2005/08/14 23:36:13  alankila
 * - set # of channels directly. What I really need is to control number of
 *   input and output channels separately, but it seems impossible with OSS.
 *   It's probably time to switch to ALSA.
 *
 * Revision 1.25  2005/08/11 17:57:22  alankila
 * - add some missing headers & fix all compiler warnings on gcc 4.0.1+ -Wall
 *
 * Revision 1.24  2005/08/08 12:03:26  fonin
 * Fixed include sys/select.h which did not work on windows.
 *
 * Revision 1.23  2005/08/07 13:13:14  alankila
 * oops: reinstate the MAX_BUFFER_SIZE / sizeof(SAMPLE) in the rdbuf & procbuf.
 * Removing that was a mistake. Keep the rest, though.
 *
 * Revision 1.22  2005/08/07 13:03:57  alankila
 * - add select() around the read() part to read and discard buffers if
 *   we begin to fall behind in audio processing
 *
 * Revision 1.21  2005/04/15 14:37:41  fonin
 * Fixed version variable
 *
 * Revision 1.20  2005/04/06 19:34:58  fonin
 * Fixed the accidental typo with "count=bits >> 8" that caused the floating exception
 *
 * Revision 1.19  2004/10/21 11:19:18  dexterus
 * Bug in the win 32 section related to sample type (SAMPLE insted of
 * DSP_SAMPLE ) fixed -- win32 working
 *
 * Revision 1.18  2004/08/10 15:07:31  fonin
 * Support processing in float/int - type DSP_SAMPLE
 *
 * Revision 1.17  2003/05/30 12:49:23  fonin
 * log2() renamed to my_log2() since log2 is a reserved word on MacOS X.
 *
 * Revision 1.16  2003/03/25 14:03:01  fonin
 * Work around buffer overruns with DirectSound playback.
 *
 * Revision 1.15  2003/03/23 20:05:42  fonin
 * New playback method via DirectSound.
 *
 * Revision 1.14  2003/03/15 20:07:01  fonin
 * Moved function expired(), fixed compilation error.
 *
 * Revision 1.13  2003/03/09 21:05:57  fonin
 * Internal redesign for new "change sampling params" feature.
 * New functions init_sound() and close_sound().
 *
 * Revision 1.12  2003/02/11 21:45:03  fonin
 * URL fixes.
 *
 * Revision 1.11  2003/02/05 21:10:10  fonin
 * Cleanup before release.
 *
 * Revision 1.10  2003/02/03 17:24:04  fonin
 * Disclaimer and legal notice text moved to a string constants to gui.h
 *
 * Revision 1.9  2003/02/03 11:39:25  fonin
 * Copyright year changed.
 *
 * Revision 1.8  2003/01/31 15:18:04  fonin
 * Few cleanups, more comments; start recording AFTER all input buffers are
 * queued.
 *
 * Revision 1.7  2003/01/30 21:33:31  fonin
 * - Added demo version code for Win32;
 * - NCHANNELS now is used in UNIX build.
 *
 * Revision 1.6  2003/01/29 19:34:00  fonin
 * Win32 port.
 *
 * Revision 1.5  2001/06/02 14:05:59  fonin
 * Added GNU disclaimer.
 *
 * Revision 1.4  2001/04/27 14:29:14  fonin
 * <sys/soundcard.h> for better compatibility with FreeBSD.
 *
 * Revision 1.3  2001/03/25 17:42:55  fonin
 * Switching back to real user identifier immediately after setting realtime priority.
 *
 * Revision 1.2  2001/03/25 12:10:06  fonin
 * Text messages begin from newline rather than end with it.
 *
 * Revision 1.1.1.1  2001/01/11 13:21:53  fonin
 * Version 0.1.0 Release 1 beta
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#ifndef _WIN32
#include <unistd.h>
#include <pthread.h>
#endif

#include "pump.h"
#include "main.h"
#include "tracker.h"
#include "gui.h"

volatile audio_driver_t  *audio_driver = NULL;
volatile int    state;
char            version[13] = "GNUitar "VERSION;

my_mutex        snd_open=NULL;
#ifndef _WIN32
pthread_t       audio_thread = 0;

SAMPLE32        wrbuf[MAX_BUFFER_SIZE * MAX_CHANNELS];
SAMPLE32        rdbuf[MAX_BUFFER_SIZE * MAX_CHANNELS];
DSP_SAMPLE_ALIGN procbuf[MAX_BUFFER_SIZE * MAX_CHANNELS];
DSP_SAMPLE_ALIGN procbuf2[MAX_BUFFER_SIZE * MAX_CHANNELS];
#else
HANDLE          audio_thread = 0;

char            wrbuf[MIN_BUFFER_SIZE * MAX_BUFFERS];
char            rdbuf[MIN_BUFFER_SIZE * MAX_BUFFERS];
DSP_SAMPLE      procbuf[MAX_BUFFER_SIZE / sizeof(SAMPLE16)];
DSP_SAMPLE      procbuf2[MAX_BUFFER_SIZE / sizeof(SAMPLE16)];
extern short    dsound;         /* from audio-windows.h */
#endif

int
main(int argc, char **argv)
{
    int             error = 0;
#ifndef _WIN32
    int             max_priority;
    struct sched_param p;

    max_priority = sched_get_priority_max(SCHED_FIFO);
    p.sched_priority = max_priority/2;

    if (sched_setscheduler(0, SCHED_FIFO, &p)) {
	fprintf(stderr, "warning: unable to set realtime priority (needs root privileges)\n");
    }

    /*
     * We were running to this point as setuid root program.
     * Switching to our native user id
     */
    setuid(getuid());
#endif

    printf(COPYRIGHT
       "This program is a free software under the GPL;\n"
       "see Help->About for details.\n");

    gtk_init(&argc, &argv);
    load_settings();

#ifndef _WIN32
    state = STATE_PAUSE;
    /* choose audio driver if not given in config */
    if (audio_driver == NULL) {
        fprintf(stderr, "Discovering audio driver.\n");
#    ifdef HAVE_JACK
        if (jack_available()) {
            audio_driver = &jack_driver;
        } else
#    endif
#    ifdef HAVE_ALSA
        if (alsa_available()) {
            audio_driver = &alsa_driver;
        } else
#    endif
#    ifdef HAVE_OSS
        if (oss_available()) {
            audio_driver = &oss_driver;
        } else
#    endif
            printf("warning: no usable audio driver found. Tried jack, alsa and oss (if compiled in.)\n");
    }
    g_thread_init(NULL);
    my_create_mutex(&snd_open);
    my_lock_mutex(snd_open);

    if (audio_driver) {
        if (pthread_create(&audio_thread, NULL, audio_driver->thread, NULL)) {
            fprintf(stderr, "Audio thread creation failed!\n");
            return ERR_THREAD;
        }
    }
#else
    state = STATE_START_PAUSE;

    if (audio_driver == NULL) {
        audio_driver = &windows_driver;
        dsound=1;
    }
    my_create_mutex(&snd_open);
    /*
     * create audio thread
     */
    audio_thread =
	CreateThread(0, 0, (LPTHREAD_START_ROUTINE) audio_driver->audio_proc, 0,
		     0, &thread_id);
    if (!audio_thread) {
	fprintf(stderr, "Can't create WAVE recording thread! -- %08X\n",
		GetLastError());
	return ERR_THREAD;
    }

    /*
     * set realtime priority to the thread
     */
    if (!SetThreadPriority(audio_thread, THREAD_PRIORITY_TIME_CRITICAL)) {
	fprintf(stderr,
		"\nFailed to set realtime priority to thread: %s. Continuing with default priority.",
		GetLastError());
    }
#endif
    pump_start(argc, argv);    

    if (audio_driver && (error = audio_driver->init()) != ERR_NOERROR) {
        fprintf(stderr, "warning: unable to begin audio processing (code %d)\n", error);
    }

    init_gui();
    gtk_main();
#ifndef _WIN32
    /* wait for audio thread to finish */
    if (state == STATE_PAUSE || state == STATE_ATHREAD_RESTART) {
        state = STATE_EXIT;
        my_unlock_mutex(snd_open);
        if (audio_thread)
            pthread_join(audio_thread, NULL);
    } else {
        state = STATE_EXIT;
        if (audio_thread)
            pthread_join(audio_thread, NULL);
        if (audio_driver)
            audio_driver->finish();
        my_unlock_mutex(snd_open);
    }
#else
    state = STATE_EXIT;
    my_unlock_mutex(snd_open);
    if (audio_driver)
        audio_driver->finish();
#endif
    pump_stop();
    save_settings();
    my_close_mutex(snd_open);

    return ERR_NOERROR;
}

